bittide/src/Bittide/Transceiver.hs

```diff
@@ -7,453 +7,556 @@
-
-  Reset freeclk ->
+  , HasSynchronousReset free
+  , HasDefinedInitialValues free
 
-  Vec chansUsed String ->
-  Vec chansUsed String ->
+  , KnownDomain rxS
+  , KnownDomain txS
+  , KnownDomain ref
+  , KnownDomain free
+  ) =>
+  Config free ->
+  Inputs tx rx ref free rxS n ->
+  Outputs n tx rx txS free
+transceiverPrbsN opts inputs@Inputs{clock, reset, refClock} = Outputs
+  -- tx
+  { txClocks          = map (.txClock) outputs
+  , txResets          = map (.txReset) outputs
+  , txReadys          = map (.txReady) outputs
+  , txSamplings       = map (.txSampling) outputs
+
+  -- rx
+  , rxClocks = map (.rxClock) outputs
+  , rxResets = map (.rxReset) outputs
+  , rxDatas  = map (.rxData) outputs
 
-  Vec chansUsed (Signal rxS (BitVector 1)) ->
-  Vec chansUsed (Signal rxS (BitVector 1)) ->
+  -- transceiver
+  , txPs = pack <$> bundle (map (.txP) outputs)
+  , txNs = pack <$> bundle (map (.txN) outputs)
 
-  TransceiverOutputs chansUsed tx rx txS freeclk
-transceiverPrbsN refclk freeclk rst chanNms clkPaths rxns rxps = TransceiverOutputs
-  { txClocks = map (.txClock) outputs
-  , rxClocks = map (.rxClock) outputs
-  , txPs     = map (.txP)     outputs
-  , txNs     = map (.txN)     outputs
-  , linkUps  = map (.linkUp)  outputs
-  , stats    = map (.stats)   outputs
+  -- free
+  , linkUps    = map (.linkUp)  outputs
+  , linkReadys = map (.linkReady)  outputs
+  , stats      = map (.stats)   outputs
   }
  where
-  outputs = zipWith4 (transceiverPrbs refclk freeclk rst) chanNms clkPaths rxns rxps
+  -- XXX: Replacing 'zipWithN' with '<$>' and '<*>' triggers a combination of:
+  --
+  --       * https://github.com/clash-lang/clash-compiler/issues/2723
+  --       * https://github.com/clash-lang/clash-compiler/issues/2722
+  --
+  outputs = zipWith8 go
+    (iterateI (+1) 0)
+    inputs.channelNames
+    inputs.clockPaths
+    (unbundle (unpack <$> inputs.rxNs))
+    (unbundle (unpack <$> inputs.rxPs))
+    inputs.txDatas
+    inputs.txReadys
+    inputs.rxReadys
+
+  go transceiverIndex channelName clockPath rxN rxP txData txReady rxReady =
+    transceiverPrbs Gth.gthCore opts Input
+      { channelName, clockPath, rxN, rxP, txData, txReady, rxReady, transceiverIndex
+      , clock, reset, refClock
+      }
 
 transceiverPrbs ::
+  forall tx rx ref free txS rxS serializedData .
   ( HasSynchronousReset tx
   , HasDefinedInitialValues tx
 
   , HasSynchronousReset rx
   , HasDefinedInitialValues rx
 
-  , HasSynchronousReset freeclk
-  , HasDefinedInitialValues freeclk
-  ) =>
+  , HasSynchronousReset free
+  , HasDefinedInitialValues free
 
-  Clock refclk ->
-  Clock freeclk ->
+  , KnownDomain rxS
+  , KnownDomain txS
+  , KnownDomain ref
+  , KnownDomain free
+  ) =>
+  GthCore tx rx ref free txS rxS serializedData ->
+  Config free ->
+  Input tx rx ref free rxS serializedData ->
+  Output tx rx txS free serializedData
+transceiverPrbs gthCore opts args@Input{clock, reset} =
+  when opts.debugIla debugIla `hwSeqX` result
```

Comment:
Have you checked that this works for `transceiverPrbsN` with `n > 1`, because I had issues with shared names in case of introducing multiple ILAs requiring the introduction of individual name sets to work around: cf. https://github.com/bittide/bittide-hardware/blob/c81b2d06ffc99bf1798421a871f61003c6f89714/bittide-instances/src/Bittide/Instances/Hitl/IlaPlot.hs#L537-L554 _(The case switch is needed here as composing names based on a type level parameter does not work with the blackbox backend at the moment.)_

Hence, if you run into the same issues here, it might be good advise to restrict the ILA debugging feature to `transceiverPrbs` only and remove it from `transceiverPrbsN`.

Reply:
Yes, it works. Though it is kind of a happy coincidence. See:

```haskell
  -- XXX: Replacing 'zipWithN' with '<$>' and '<*>' triggers a combination of:
  --
  --       * https://github.com/clash-lang/clash-compiler/issues/2723
  --       * https://github.com/clash-lang/clash-compiler/issues/2722
  --
```

Reply:
Can you please add another note to the code then. The above comment only legitimates the use of `zipWithN` instead of Applicative primitives, but does not mention that the debug ILAs start breaking as well, if not used or the two mentioned issues get fixed. 

We need to inform the person removing that `XXX` note some time in the future, that the ILAs need to be fixed as well as part of the change introduced then. Let's be some kind people of past :wink:.